channeldb/mp_payment.go

```diff
@@ -267,19 +294,171 @@ func (m *MPPayment) Registrable() error {
 	// There are still inflight HTLCs and we need to check whether there
 	// are settled HTLCs or the payment is failed. If we already have
 	// settled HTLCs, we won't allow adding more HTLCs.
-	if settled {
+	if m.State.HasSettledHTLC {
 		return ErrPaymentPendingSettled
 	}
 
 	// If the payment is already failed, we won't allow adding more HTLCs.
-	if failed {
+	if m.State.PaymentFailed {
 		return ErrPaymentPendingFailed
 	}
 
 	// Otherwise we can add more HTLCs.
 	return nil
 }
 
+// setState creates and attaches a new MPPaymentState to the payment. It also
+// updates the payment's status based on its current state.
+func (m *MPPayment) setState() error {
+	// Fetch the total amount and fees that has already been sent in
+	// settled and still in-flight shards.
+	sentAmt, fees := m.SentAmt()
+
+	// Sanity check we haven't sent a value larger than the payment amount.
+	totalAmt := m.Info.Value
+	if sentAmt > totalAmt {
+		return fmt.Errorf("%w: sent=%v, total=%v", ErrSentExceedsTotal,
+			sentAmt, totalAmt)
+	}
+
+	// Get any terminal info for this payment.
+	settle, failure := m.TerminalInfo()
+
+	// Now determine the payment's status.
+	status, err := decidePaymentStatus(m.HTLCs, m.FailureReason)
+	if err != nil {
+		return err
+	}
+
+	// Update the payment state and status.
+	m.State = &MPPaymentState{
+		NumAttemptsInFlight: len(m.InFlightHTLCs()),
+		RemainingAmt:        totalAmt - sentAmt,
+		FeesPaid:            fees,
+		HasSettledHTLC:      settle != nil,
+		PaymentFailed:       failure != nil,
+	}
+	m.Status = status
+
+	return nil
+}
+
+// SetState calls the internal method setState. This is a temporary method
+// to be used by the tests in routing. Once the tests are updated to use mocks,
+// this method can be removed.
+//
+// TODO(yy): delete.
+func (m *MPPayment) SetState() error {
+	return m.setState()
+}
+
+// NeedWaitAttempts decides whether we need to hold creating more HTLC attempts
+// and wait for the results of the payment's inflight HTLCs. Return an error if
+// the payment is in an unexpected state.
+func (m *MPPayment) NeedWaitAttempts() (bool, error) {
+	// Check when the remainingAmt is not zero, which means we have more
+	// money to be sent.
+	if m.State.RemainingAmt != 0 {
+		switch m.Status {
+		// If the payment is newly created, no need to wait for HTLC
+		// results.
+		case StatusInitiated:
+			return false, nil
+
+		// If we have inflight HTLCs, we'll check if we have terminal
+		// states to decide if we need to wait.
+		case StatusInFlight:
+			// We still have money to send, yet one of the HTLCs is
+			// settled, return an error as the peer is violating
+			// the protocol.
+			if m.State.HasSettledHTLC {
+				return false, fmt.Errorf("%w: settled HTLC "+
```

Comment:
Wouldn't it make sense to also wait in this case to collect all results from the still inflight HTLCs and only log the inconsistency? That way we would update MC as in the failed payment case later. Although I think this code is not even reachable because it's blocked by `AllowMoreAttempts` and the `Registrable` check in there.

Reply:
was thinking about fixing the stuck inflight payments, but I think you are right, changed to wait for the attempts. In a follow-up PR, I will check the case where one of the HTLCs is timed out. If that happens, we'll also update the payment, in which case the sender will get a free HTLC.

Reply:
is that follow-up PR already open? the comment would also need an update

Reply:
updated the comments. Nope no open PR yet.